### ISSUE TITLE
[WICKET-7023] Problem when comparing model instead of object in NestedTree class 

### ISSUE THREAD

**Issue by jira-importer** [jira-importer/jira-to-github-issues#147]
<p>In NestedTree abstract class (and Maybe in other components or classes) we use the model comparaison instead of object comparaison (model.getObject()).</p>

<p>In IModel or in LoadableDetachableModel We don't have any equals method defined (no default, nothing), it will use the default one what is not good.</p>

<p>example (l111) of NestedTree.java in wicket-extensions library</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">
@Override
    <span class="code-keyword">public</span> void updateBranch(T t, IPartialPageRequestHandler target)
    {
        <span class="code-keyword">final</span> IModel<T> model = getProvider().model(t);
        visitChildren(BranchItem.class, <span class="code-keyword">new</span> IVisitor<BranchItem<T>, <span class="code-object">Void</span>>()
        {
            @Override
            <span class="code-keyword">public</span> void component(BranchItem<T> branch, IVisit<<span class="code-object">Void</span>> visit)
            {
                <span class="code-keyword">if</span> (model.equals(branch.getModel()))
                {
                    <span class="code-comment">// BranchItem always outputs its markupId
</span>                    target.add(branch);
                    visit.stop();
                }
            }
        });
        model.detach();
    }</pre>
</div></div>
<p>{{}}<br/>
Instead of comparing <em><b>model.equals(branch.getModel()</b></em> should we use instead the object comparaison itself :</p>

<p>Objects.equals<b><em>( model.getObject(), branch.getModelObject() )</em></b></p>

<p><em>because here model are different and not same ref , it will always return false ? or i miss something ?</em></p>

<p>If not could you explain us why we have to use the model and what to do to make it work properly ?</p>

<p>Even we use a custom extension of LoadableDetachableModel if we override the model equals we have to check the state, if state is attached, get the transient object and compare it, if not attached, compare the unid. Moreover it is possible in wicket core or other librairies to compare 2 models not attached (so finally quite ugly equals no ?).</p>

<p>What is the default behaviour of comparing models if we used the default models in wicket ? because even in wicket models (CompoundPropertyModel for instance no equals is defined)</p>

<p> </p>

---
<details><summary><i>Originally reported by <a title="leclerc cyril" href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=abuzouzou">abuzouzou</a>, imported from: <a href="https://issues.apache.org/jira/browse/WICKET-7023" target="_blank">Problem when comparing model instead of object in NestedTree class </a></i></summary>
<i><ul>
<li><b>assignee</b>: <a title="Sven Meier" href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=svenmeier">svenmeier</a>
<li><b>status</b>: Closed
<li><b>priority</b>: Major
<li><b>resolution</b>: Not A Problem
<li><b>resolved</b>: 2023-01-11T14:27:30+00:00
<li><b>imported</b>: 2025-11-27
</ul></i>
</details>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=JIRAUSER290330">JIRAUSER290330</a>:</i>
<p>Hello, any comment on this ? real issue, bad using of NestedTree etc ?</p>

<p>Thanks you very much</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=svenmeier">svenmeier</a>:</i>
<p>Please read the javadoc of ITreeProvider#model(T):</p>

<p>"Important note: The model must implement {@link Object#equals(Object)} and {@link Object#hashCode()} !"</p>

<p>This implementation was chosen deliberately to prevent unnecessary calls to #getObject().</p>

<p>Please use a custom implementation, if not default does it for you. Many projects have something like JpaEntityModel. See org.apache.wicket.examples.tree.FooProvider/FooMode for inspiration.</p>

<p>And please use the mailing list for questions the next time.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=JIRAUSER290330">JIRAUSER290330</a>:</i>
<p>Ok thank you, but just one question, why any model from wicket implement hashcode and equals ?</p>

<p>it seems FooProvider and FooMode not existing anymore on version 9 wicket (last one seems to be version 8)</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=svenmeier">svenmeier</a>:</i>
<p>Sorry, but I don't understand your question.</p>

<p>The classes are still in wicket-examples:<br/>
<a href="https://github.com/apache/wicket/blob/master/wicket-examples/src/main/java/org/apache/wicket/examples/tree/FooProvider.java" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/wicket/blob/master/wicket-examples/src/main/java/org/apache/wicket/examples/tree/FooProvider.java</a></p>
